### PR TITLE
Mimir query engine: add HTTP header to force use of Prometheus' engine

### PR DESCRIPTION
**Pull request by charleskorn**
#### What this PR does

This PR adds support for forcing the use of Prometheus' engine when the Mimir query engine is enabled by sending the `X-Mimir-Force-Prometheus-Engine: true` HTTP header with the query request.

The header is only effective if fallback to Prometheus' engine is enabled and fallback is enabled (ie. `-querier.enable-promql-engine-fallback=true`).

#### Which issue(s) this PR fixes or relates to

(none)

#### Checklist

- [x] Tests updated.
- [n/a] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [n/a] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/streamingpromql/compat/fallback_engine.go [hunk 1]
```diff
@@ -27,6 +27,8 @@ type EngineWithFallback struct {
 	logger log.Logger
 }
 
+var errFallbackForcedByHTTPHeader = NotSupportedError{"fallback forced by HTTP header"}
```
Comment (jhesketh) [inline pkg/streamingpromql/compat/fallback_engine.go:30]: (nit)
I don't know if I like using `NotSupportedError` here. It's not that the query is necessarily unsupported, but that we aren't going to use the query engine. Maybe we need an error class "MQEError" or something that NotSupportError and Fallback use?

Similarly renaming `e.unsupportedQueries` to FallbackQueries or similar.

File: pkg/streamingpromql/compat/fallback_header_test.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestEngineFallbackInjector(t *testing.T) {
+	testCases := map[string]struct {
+		headers http.Header
+
+		forceFallback bool
```
Comment (jhesketh) [inline pkg/streamingpromql/compat/fallback_header_test.go:17]: (nit)
```suggestion
		expectedForceFallback bool
```

**Review by jhesketh — APPROVED** [review]

File: pkg/streamingpromql/compat/fallback_header_test.go (hunk 2 above)
Comment (charleskorn) [reply pkg/streamingpromql/compat/fallback_header_test.go]: Resolved in ef55390

File: pkg/streamingpromql/compat/fallback_engine.go (hunk 1 above)
Comment (charleskorn) [reply pkg/streamingpromql/compat/fallback_engine.go]: Good point, I've dropped the awkward use of `NotSupportedError` in 28c4bc2.

I've kept `unsupportedQueries` as-is for now - if it annoys us in the future, we can change it.

**Pull request merged**